tokio/src/runtime/basic_scheduler.rs

```diff
@@ -214,12 +282,22 @@ where
     CURRENT.set(context, || f(scheduler, context))
 }
 
-impl<P> Drop for BasicScheduler<P>
-where
-    P: Park,
-{
+impl<P: Park> Drop for BasicScheduler<P> {
     fn drop(&mut self) {
-        enter(self, |scheduler, context| {
+        // Avoid a double panic if we are currently panicking and
+        // the lock may be poisoned.
+        let mut inner = if let Ok(lock) = &mut self.inner.lock() {
+            lock.take()
+                .expect("Oh no! We never placed the Inner state back, this is a bug!")
+        } else {
+            if std::thread::panicking() {
+                return;
+            } else {
+                panic!("Inner lock poisoned");
+            }
+        };
```

Comment:
i really think we just want to always put the state back even if the mutex is already poisoned, like this:
```suggestion
        let mut inner = { 
            let lock = self.inner.lock().unwrap_or_else(|e| e.into_inner());
            match lock.take() {
                Some(inner) => inner,
                None if std::thread::panicking() => return,
                None => panic!("Oh no! We never placed the Inner state back, this is a bug!"),
            }
        };
```